### DISCUSSION TITLE
Retrive records [RESOLVED]

### DISCUSSION THREAD

**Discussion by Giovanni-Petrella** [filamentphp/filament#5144] [Help]
How can I retrieve the records of a table in a custom page?

`class ReceiptsSummary extends Page implements Forms\Contracts\HasForms, HasTable
{
    use Forms\Concerns\InteractsWithForms;
    use InteractsWithTable;

    public $year;
    public $season;
    public $type;
    public $athlete;

    protected static ?string $navigationIcon = 'heroicon-o-document-text';

    protected static string $view = 'filament.pages.receipts-summary';

    protected static ?string $navigationGroup = 'Reports';

    protected static string $resource = ReceiptResource::class;

    protected static ?string $model = Receipt::class;

    protected static function getNavigationGroup(): ?string
    {
        return trans('global.menu_reports');
    }

    protected static function getNavigationLabel(): string
    {
        return trans('receipts_summary.title_plural');
    }

    protected function getTitle(): string
    {
        return trans('receipts_summary.title_plural');
    }

    public function mount(Receipt $record)
    {
        $this->record = $record;

        $this->receiptsForm->fill([
            'year' => Carbon::now()->year,
        ]);
    }

    protected function getForms(): array
    {
        return [
            'receiptsForm' => $this->makeForm()
                ->schema($this->getReceiptsFormSchema())
                ->model($this->record),
        ];
    }

    public function filter()
    {
        $data = $this->receiptsForm->getState();
        $this->year = $data['year'] ?? null;
        $this->season = $data['season'] ?? null;
        $this->type = $data['type'];
        $this->athlete = $data['athlete'];
    }

    protected function getReceiptsFormSchema(): array
    {
        return [
            Placeholder::make('')
                ->content(trans('receipts_summary.filter_receipts')),
            Card::make()
                ->schema([
                    Radio::make('type')
                        ->label(trans('receipts_summary.type'))
                        ->options([
                            'season' => trans('receipts_summary.season'),
                            'year' => trans('receipts_summary.year'),
                        ])
                        ->descriptions([
                            'season' => trans('receipts_summary.type_is_season'),
                            'year' => trans('receipts_summary.type_is_year'),
                        ])
                        ->reactive()
                        ->required()
                        ->columnSpan(1),
                    Select::make('season')
                        ->label(trans('receipts_summary.season'))
                        ->required()
                        ->hidden(fn (Closure $get) => $get('type') !== 'season')
                        ->hint(trans('receipts_summary.season_hint'))
                        ->options(Season::where('club_id', auth()->user()->club_id)->pluck('season', 'id')),
                    TextInput::make('year')
                        ->label(trans('receipts_summary.year'))
                        ->numeric()
                        ->length(4)
                        ->minValue(2000)
                        ->maxValue(2100)
                        ->prefixIcon('heroicon-s-calendar')
                        ->hint(trans('receipts_summary.year_hint'))
                        ->hidden(fn (Closure $get) => $get('type') !== 'year')
                        ->required(),
                    Select::make('athlete')
                        ->label(trans('receipts_summary.athlete'))
                        ->required()
                        ->hint(trans('receipts_summary.athlete_hint'))
                        ->options(Athlete::where('club_id', auth()->user()->club_id)->pluck('full_name', 'id')),
                ])
                ->columns(3),
        ];
    }

    protected function getTableQuery(): Builder
    {
        $query = Receipt::query()->where('club_id', auth()->user()->club_id)->where('subject_id', null)->where('athlete_id', $this->athlete)->ds();
        if ($this->type == 'year') {
            $query = $query->whereYear('receipt_date', $this->year);
        } elseif ($this->type == 'year') {
            $season = Season::where('season', $this->season)->first();
            $query = $query->whereBetween('receipt_date', [$season->date_start_season, $season->date_end_season]);
        }
        return $query;
    }

    public function getTableColumns(): array
    {
        return [
            TextColumn::make('receipt_number')->label(trans('receipts.receipt_number'))->sortable()->searchable()->weight('bold')->alignment('right'),
            TextColumn::make('athlete.full_name')->label(trans('receipts.receipt_athlete'))->sortable()->searchable(),
            TextColumn::make('receipt_date')->label(trans('receipts.receipt_date'))->sortable()->searchable()
                ->date('d/m/Y')->weight('bold')->alignment('right'),
            TextColumn::make('total_amount')->label(trans('receipts.receipt_total_amount'))->money('eur', true)->weight('bold')->alignment('right'),
        ];
    }

    protected function getActions(): array
    {
        return [
            Action::make('settings')
                ->label('Settings')
                // ->mountUsing(function (Collection $records) {
                //     dsd($records);
                // })
                ->action(function (Collection $records) {
                    dsd($this->actions);
                }),
        ];
    }`

**Comment by ArtMin96** [comment]
What do you mean by retrieving the records of a table?

**Comment by falconsmilie** [comment]
Not sure what you mean by your question, but i did notice the `if` and `elseif` conditions in `getTableQuery` are the same.

**Reply by Giovanni-Petrella** [reply]
thank you i resolve it changed all code ex:
`   protected function getTableQuery(): Builder
    {
        if ($this->data) {
            $query = Accounting::query()->where('club_id', auth()->user()->club_id)->where($this->reference, $this->referent)->where('expense_id', null);
            if ($this->type == 'year') {
                $query = $query->whereYear('transaction_date', $this->year);
            } elseif ($this->type == 'season') {
                $season = Season::where('season', $this->season)->first();
                if ($season) {
                    $query = $query->whereBetween('transaction_date', [$season->date_start_season, $season->date_end_season]);
                }
            }
        } else {
            $query = Accounting::query()->where('club_id', 0);
        }
        return $query;
    }`

**Comment by naumantoor** [comment]
how to did you get it in the form ?
